src/file_pipe_log/pipe.rs

```diff
@@ -18,13 +18,76 @@ use crate::metrics::*;
 use crate::pipe_log::{FileBlockHandle, FileId, FileSeq, LogQueue, PipeLog};
 use crate::{Error, Result};
 
-use super::format::{FileNameExt, Version};
+use super::format::{FileNameExt, LogFileContext, Version};
 use super::log_file::{build_file_reader, build_file_writer, FileHandler, LogFileWriter};
 
 struct FileCollection<F: FileSystem> {
+    /// first seq of active files
     first_seq: FileSeq,
+    /// final seq of active files
     active_seq: FileSeq,
+    /// first seq of stale files
+    first_stale_seq: FileSeq,
     fds: VecDeque<FileHandler<F>>,
+    /// `None` => no capbility for recycling stale files
+    /// `Some(0)` => infinite volume for recycling stale files
+    /// `Some(_)` => finite volume for recycling stale files
+    capacity: Option<usize>,
```

Comment:
No need for option. There should always be a limit to how many files we keep around. When capacity==0, recycling is naturally disabled.
```suggestion
    /// Try to reserve and recycle obsolete files unless total file count exceeds capacity.
    capacity: usize,
```